helper/progress/chain.go

```diff
@@ -0,0 +1,125 @@
+package progress
+
+import (
+	"sync"
+
+	"github.com/0xPolygon/polygon-sdk/blockchain"
+)
+
+type ChainSyncType string
+
+const (
+	ChainSyncRestore ChainSyncType = "restore"
+	ChainSyncBulk    ChainSyncType = "bulk-sync"
+)
+
+// Progression defines the status of the sync
+// progression of the node
+type Progression struct {
+	// SyncType is indicating the sync method
+	SyncType ChainSyncType
+
+	// StartingBlock is the initial block that the node is starting
+	// the sync from. It is reset after every sync batch
+	StartingBlock uint64
+
+	// CurrentBlock is the last written block from the sync batch
+	CurrentBlock uint64
+
+	// HighestBlock is the target block in the sync batch
+	HighestBlock uint64
+}
+
+type ProgressionWrapper struct {
+	// progression is a reference to the ongoing batch sync.
+	// Nil if no batch sync is currently in progress
+	progression *Progression
+
+	// stopCh is the channel for receiving stop signals
+	// in progression tracking
+	stopCh chan struct{}
+
+	lock sync.RWMutex
+
+	syncType ChainSyncType
+}
+
+func NewProgressionWrapper(syncType ChainSyncType) *ProgressionWrapper {
+	return &ProgressionWrapper{
+		progression: nil,
+		stopCh:      make(chan struct{}),
+		syncType:    syncType,
+	}
+}
+
+// startProgression initializes the progression tracking
+func (pw *ProgressionWrapper) StartProgression(
+	startingBlock uint64,
+	subscription blockchain.Subscription,
+) {
+	pw.lock.Lock()
+	defer pw.lock.Unlock()
+
+	pw.progression = &Progression{
+		SyncType:      pw.syncType,
+		StartingBlock: startingBlock,
+	}
+
+	go pw.RunUpdateLoop(subscription)
+}
+
+// runUpdateLoop starts the blockchain event monitoring loop and
+// updates the currently written block in the batch sync
+func (pw *ProgressionWrapper) RunUpdateLoop(subscription blockchain.Subscription) {
+	eventCh := subscription.GetEventCh()
+	for {
+		select {
+		case event := <-eventCh:
+			if event.Type == blockchain.EventFork {
+				continue
+			}
+
+			if len(event.NewChain) == 0 {
+				continue
+			}
+
+			pw.UpdateCurrentProgression(event.NewChain[0].Number)
```

Comment:
I'm just noticing it now, do you think it is incorrect to assume that NewChain has a single block number?

Reply:
@zivkovicmilos You're right. I fixed at a5607d0e08a12df4ef44a5949d39ceb63a134b61